### ISSUE TITLE
Make tenants configurable from the host application

### ISSUE THREAD

**Issue by sebastienros** [OrchardCMS/OrchardCore#1755]
Example

```
services.AddOrchardCore()
    .AddModules(options =>
    {
        options.RegisterTenantServices(collection =>
            collection.AddDataProtection()
                .SetApplicationName("my-application-name")
                .PersistKeysToAwsS3(Configuration.GetSection("myS3XmlStorageConfiguration")));
    });
```

We need to be able to configure services, and middleware. Also the configuration should be able to take the tenant information into account (name, App_data folder).

This should have a higher priority than modules, as code in the host is always custom while features might be enabled by some code that is not accessible from the deb.

**Comment by jtkech** [comment]
Another example at the host level in `AddOrchardCms()` (removed in the `OC.DataProtection` module).

    services // Or e.g AddOrchardCms()
        .ConfigureTenantServices<IDataProtectionProvider, IOptions<ShellOptions>, ShellSettings>(
            (collection, provider, options, settings) =>
        {
            var directory = Directory.CreateDirectory(Path.Combine(
                options.Value.ShellsApplicationDataPath,
                options.Value.ShellsContainerName,
                settings.Name, "DataProtection-Keys"));

            collection.Add(new ServiceCollection()
                .AddDataProtection()
                .PersistKeysToFileSystem(directory)
                .SetApplicationName(settings.Name)
                .Services);
        });

Then, not sure it works as expected because i'm a newbie in security. But at least it builds, it runs without failing and i could login to the admin.

**Comment by jtkech** [comment]
As a side note.

             // When we deine the default behavior of a tenant
            // 1. Here the service collection instance is the same as below (see 2.).
            services.AddModules(modules =>
            {
                modules.WithDefaultFeatures(
                    "OrchardCore.Antiforgery", "OrchardCore.Mvc", "OrchardCore.Settings",
                    "OrchardCore.Setup", "OrchardCore.Recipes", "OrchardCore.Commons");
            });

        // Which calls this extension
        public static ModularServiceCollection WithDefaultFeatures(
            this ModularServiceCollection modules, params string[] featureIds)
        {
            modules.Configure(
                // 2. Here we retrieve the same service collection instance as above (see 1.).
                services =>
            {
                foreach (var featureId in featureIds)
                {
                    services.AddTransient(sp => new ShellFeature(featureId));
                };
            });

            return modules;
        }

So not sure that the `ModularServiceCollection` class and its extensions are useful, and because it is not a regular service collection it is less flexible. But not sure.

**Comment by jtkech** [comment]
Just tried the above, so simplifying `AddModules()` and e.g also `WithDefaultFeatures()`.

        public static IServiceCollection WithDefaultFeaturesTest(
            this IServiceCollection services, params string[] featureIds)
        {
            foreach (var featureId in featureIds)
            {
                services.AddTransient(sp => new ShellFeature(featureId));
            }

            return services;
        }

It seems to work, i could re-run a setup, by using this.

            services.AddModulesTest(collection =>
            {
                collection.WithDefaultFeaturesTest(
                    "OrchardCore.Antiforgery", "OrchardCore.Mvc", "OrchardCore.Settings",
                    "OrchardCore.Setup", "OrchardCore.Recipes", "OrchardCore.Commons");
            });

Where `collection` is a regular service collection, so that it is easier to use standard DI helpers and / or custom ones. E.g, here we could also use the new helper (see 2 comments above) like this.

            services.AddModulesTest(collection =>
            {
                collection
                    .WithDefaultFeaturesTest( ... )
                    .ConfigureTenantServices<IOptions<ShellOptions>, ShellSettings>(
                        (featureAwareServiceCollection, options, settings) =>
                    {

It was for testing, right now i will revert this change about removing `ModularServiceCollection`.

**Comment by sebastienros** [comment]
I like it, and being able to enable specific features by default in the host application is much better that having the specific Commons modules.

**Comment by jtkech** [comment]
Okay cool.

We could also remove e.g `OC.DataProtection` and potentially all modules which are only there to execute their startup methods. So maybe also `OC.Mvc` but not sure it's good to do this.

About ordering, the new helpers create and register an underlying `Modules.Startup` with the hard coded order of 10000 when related to configure services so that they can replace services, and -100000 when related to configure the pipeline so that e.g adding a middleware will run before.

E.g as i remember `OC.Mvc` needs to run very first (it has a priority of -1 to do this), so if it is to configure services (not the pipeline) the host helper would not work, i will try it.

Hmm, maybe good with the new helpers to be able to change the order value. And maybe good that `StartupBase` has 2 `Order` properties, one for services, one for the pipeline, but not sure, i will see.

Note: just get rid of the `ModularServiceCollection`, simpler and seems to work.

**Comment by jtkech** [comment]
@sebastienros no problem but because there are many possibilities i need some orientations.

So, i could replace the `OrchardCore.Mvc` module by just doing this at the host level.

    services.AddModules()
        ...
        .ConfigureTenantServices<IServiceProvider>((collection, sp) =>
        {
            collection.AddMvcModules(sp);
        })

        .ConfigureTenant((app, routes, sp) =>
         {
             app.UseStaticFilesModules();
         });

**Note**: This could be wrapped in only one e.g `AddMvcModules()` called at the host level.

I could do this because the new helpers register transient instances at the host level but which are resolved and executed in a tenant context. It is inspired on how aspnet implement options configuration.

**Note**: When an instance has no injected dependencies, aspnet register a singleton. But here i still used a transient lifetime (that's ok) because when we have multiple singletons of the same type registered at the host level, there is an issue when we clone them in our child container, this could be fixed by #1644.

---

So, it's like having a `Module` defined at the host level and then enabled for each tenant. That said, right now this module only have startup(s) and no features.

That's why i could only replace a part of the `"OrchardCore.Commons"` module.

        // Replace part of "OrchardCore.Commons"
        .ConfigureTenantServices(collection =>
        {
            collection.AddDeferredTasks();
            collection.AddDataAccess();
            collection.AddBackgroundTasks();
            collection.AddResourceManagement();
            collection.AddCaching();
            collection.AddShellDescriptorStorage();
            collection.AddExtensionManager();

            // Can't be done here because it defines shapes that need to be tied to a feature
            //collection.AddTheming();

            collection.AddLiquidViews();
        })

        .ConfigureTenant((app, routes, sp) =>
        {
            sp.AddTagHelpers(typeof(ResourcesTagHelper).Assembly);
            sp.AddTagHelpers(typeof(ShapeTagHelper).Assembly);
        })

- So, is it good to replace some modules? If so which ones knowing that we will not be able to enable / disable them. Some candidates are `".Antiforgery", ".DataProtection" ".Mvc", ".Commons"`. But maybe always better to add some functionalities by just referencing a module.

 Note: `OC.Commons` also define a filter, a controller and a view used when we are using the saas recipe.

- Is it good to be able to define at the host level a more complete module, startup(s) and at least one feature? So that maybe we could replace completely `OC.Commons`.

- Hmm, maybe good that the application, becoming a module (enabled for all tenants), could define some controllers, views, and / or razor pages, but not sure it is doable, and worth to try.

- Hmm, if we can define a Host module i think we will also be able to enable / disable its feature(s), but not sure it is good here if there are core features, but maybe it would depend.

**Comment by PinpointTownes** [comment]
Glad to see that you guys like my idea :clap:  

(that sounds like a good opportunity to also implement https://github.com/OrchardCMS/OrchardCore/issues/1105, assuming you like the syntax).
